### DISCUSSION TITLE
How to create an image with a user and a script?

### DISCUSSION THREAD

**Discussion by paulmenzel** [systemd/mkosi#2617] [Q&A]
I’d like to create an image and install [BigBlueButton 3.0](https://docs.bigbluebutton.org/3.0/administration/install/) in it. That requires Ubuntu 22.04 (*jammy*) and a user *bbb*, that is in group *sudo*, and then runs the script [`bbb-install.sh`](https://github.com/bigbluebutton/bbb-install/) .

The script creates Let’s Encrypt certificates and checks certain thinks in the end, so needs to be run in the VM with network access. So I’d only like to have it present in the image.

```
$ mkdir ~/src/ubuntu-bbb
$ cd ~/src/ubuntu-bbb
$ mkdir mkosi.output
$ more mkosi.local.conf
[Distribution]
Repositories=main,universe

[Content]
Packages=dbus
#         linux-image-kvm
         systemd
         udev
         wget
         etckeeper
Autologin=no
WithNetwork=true
$ more mkosi.prepare
set -x
env
USER=bbb

useradd --create-home --user-group $USER --password '$6$jQZgb5KcbbClbLp3$mWcy2jgqjhpBSU1OWY3A77M28YYjhEZSfkXRlmnFpSnCMObDzuh8aAjgbI70eYjsI70i55Oro5WlSCx0OcFYI.'
apt install sudo
adduser bbb sudo
cd /home/bbb
wget https://github.com/paulmenzel/bbb-install/raw/install-docker-and-docker-compose-from-ubuntu-archives/bbb-install.sh
chown bbb: bbb-install.sh
```

Then

    mkosi -d ubuntu -r jammy -t disk -o ubuntu-jammy-bigbluebutton --debug

but the user is *not* created.

```
[…]
‣  Running prepare script /home/build/src/ubuntu-bbb/mkosi.prepare…
‣ + /work/prepare final
+ env
SCRIPT=/work/prepare
PACKAGEDIR=/work/packages
HOME=/
MKOSI_CONFIG=/work/config.json
GIT_PROXY_SSL_CAINFO=/proxy.cacert
MKOSI_GID=65436
WITH_NETWORK=1
TERM=screen
CHROOT_SRCDIR=/work/src
PATH=/scripts:/usr/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
WITH_TESTS=1
LANG=C.UTF-8
DISTRIBUTION=ubuntu
ARCHITECTURE=x86-64
SRCDIR=/work/src
MKOSI_UID=65436
RELEASE=jammy
BUILDROOT=/buildroot
CHROOT_SCRIPT=/work/prepare
SYSTEMD_TMPFILES_FORCE_SUBVOL=0
WITH_DOCS=1
PWD=/work/src
SYSTEMD_OFFLINE=1
+ USER=bbb
+ useradd --create-home --user-group bbb --password $6$LFniepaPBgrQ5GAc$3D607YE3YpSQjuu9SOWH3Mv0PNCHNNihzfED/xryq6HlHJpttkQdhvbXWqaf.XJx//jWL6J12W6EEDpHgm3i4.
+ apt install sudo
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  sudo
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 821 kB of archives.
After this operation, 2568 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 sudo amd64 1.9.9-1ubuntu2.4 [821 kB]
Fetched 821 kB in 6s (142 kB/s)
Selecting previously unselected package sudo.
(Reading database ... 8667 files and directories currently installed.)
Preparing to unpack .../sudo_1.9.9-1ubuntu2.4_amd64.deb ...
Unpacking sudo (1.9.9-1ubuntu2.4) ...
Setting up sudo (1.9.9-1ubuntu2.4) ...   
Processing triggers for libc-bin (2.35-0ubuntu3.6) ...
+ adduser bbb sudo
fatal: The user `bbb' does not exist.
+ cd /home/bbb
/work/prepare: 8: cd: can't cd to /home/bbb
+ wget https://github.com/paulmenzel/bbb-install/raw/install-docker-and-docker-compose-from-ubuntu-archives/bbb-install.sh
/work/prepare: 9: wget: not found
+ chown bbb: bbb-install.sh
chown: invalid spec: ‘bbb:’
‣ "sh -c 'trap '"'"'rm -rf /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/mkosi-var-tmp-267e05f841e7414b'"'"' EXIT && mkdir --mode 1777 /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/mkosi-var-tmp-267e05f841e7414b && $0 "$@"' bwrap --die-with-parent --proc /proc --setenv SYSTEMD_OFFLINE 1 --unsetenv TMPDIR --tmpfs /tmp --unshare-ipc --dev /dev --symlink usr/bin /bin --symlink usr/sbin /sbin --symlink usr/lib /lib --symlink usr/lib64 /lib64 --setenv PATH /scripts:/usr/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games --uid 0 --gid 0 --cap-add ALL --dir /work/src --chdir /work/src --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/root /buildroot --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/pkgmngr/etc /etc --ro-bind /etc/alternatives /etc/alternatives --ro-bind /etc/ca-certificates /etc/ca-certificates --bind /etc/resolv.conf /etc/resolv.conf --ro-bind /etc/ssl /etc/ssl --ro-bind /etc/ssl/certs/ca-certificates.crt /proxy.cacert --ro-bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/mkosi-scripts2j5r023x /scripts --ro-bind /usr /usr --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/root/var/cache/apt /var/cache/apt --bind '/home/build/.cache/mkosi/ubuntu~jammy~x86-64/cache/apt/archives' /var/cache/apt/archives --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/root/var/lib/apt /var/lib/apt --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/pkgmngr/var/log /var/log --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/mkosi-var-tmp-267e05f841e7414b /var/tmp --ro-bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/tmp15lw8prm /work/config.json --bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/packages /work/packages --ro-bind /home/build/src/ubuntu-bbb/mkosi.prepare /work/prepare --ro-bind /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg/tmp/mkosi-scripts8rs15fyv /work/scripts --bind /home/build/src/ubuntu-bbb /work/src sh -c 'chmod 1777 /dev/shm && chmod 755 /etc && exec $0 "$@"' /work/prepare final" returned non-zero exit code 1.
‣ + rm -rf -- /home/build/.cache/mkosi/mkosi-workspacet7c_qbpg
‣  (Fixing ownership of package manager cache directory)
‣ + chown --recursive 65436:65436 '/home/build/.cache/mkosi/ubuntu~jammy~x86-64/cache/apt'
‣ + chown --recursive 65436:65436 '/home/build/.cache/mkosi/ubuntu~jammy~x86-64/lib/apt'
Traceback (most recent call last):
  File "/home/build/src/mkosi/mkosi/run.py", line 58, in uncaught_exception_handler
    yield
  File "/home/build/src/mkosi/mkosi/run.py", line 99, in fork_and_wait
    target(*args, **kwargs)
  File "/home/build/src/mkosi/mkosi/__init__.py", line 4378, in run_build
    build_image(Context(args, config, workspace=workspace, resources=resources))
  File "/home/build/src/mkosi/mkosi/__init__.py", line 3594, in build_image
    run_prepare_scripts(context, build=False)
  File "/home/build/src/mkosi/mkosi/__init__.py", line 585, in run_prepare_scripts
    run(
  File "/home/build/src/mkosi/mkosi/run.py", line 145, in run
    with spawn(
  File "/usr/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/home/build/src/mkosi/mkosi/run.py", line 307, in spawn
    raise subprocess.CalledProcessError(returncode, cmdline)
subprocess.CalledProcessError: Command '['/work/prepare', 'final']' returned non-zero exit status 1.
‣ + tput cnorm
‣ + tput smam
```

PS: Is there a way to “live” debug a failure, that means, to get dropped in a shell in that environment, where it fails?

PPS: I am sorry for all my question. I have the feeling I am missing some fundamental understanding about mkosi. I really feel stupid and appreciated all your support.

**Comment by behrmann** [comment]
> re: creating a user

Scripts run from the build host and not in the image by default, so you will need to use either `mkosi-chroot` or add a `.chroot` extension to the script or use `--root` options of commands when they have them or do it by hand.

Details are n the *Scripts* section of the man page.

That being said, instead of calling adduser yourself have a look at the `sysusers.d` man page.

> re: network access

see the `WithNetwork=` option for `mkosi.build` scripts

> PS: Is there a way to “live” debug a failure, that means, to get dropped in a shell in that environment, where it fails?

See the options `--debug`, `--debug-shell` and `--debug-workspace`. The middle one is probably what you are looking for.

> PPS: I am sorry for all my question. I have the feeling I am missing some fundamental understanding about mkosi. I really feel stupid and appreciated all your support.

No worries. `mkosi documentation` will show you the man page of the version you are using.

**Reply by paulmenzel** [reply]
Thank you. I hit some more issues, and would keep discussing this in here. If that’s unwanted, please tell me.

**Comment by paulmenzel** [comment]
What package should I install in Ubuntu 22.04 to get the init system working? With

    [Distribution]
    Repositories=main,universe

    [Content]
    Packages=dbus
             linux-image-kvm
             systemd
             systemd-repart
             udev
             wget
             etckeeper
             ca-certificates
             grub-pc
             bash
             apt
             sudo
             ssh
    Autologin=no
    WithNetwork=true

GRUB does not seem to be correctly installed, and

    qemu-system-x86_64 -M q35 -m 32G -enable-kvm -cpu host -smp cpus=32 -device virtio-rng-pci -net nic,model=virtio-net-pci -net user,hostfwd=tcp::22231-:22 -drive if=virtio,file=/scratch/local2/ubuntu-jammy-bigbluebutton.qcow2 -nographic

does not start anything.

Using the Linux kernel image directly, the init system fails:

    $ qemu-system-x86_64 -M q35 -m 32G -enable-kvm -cpu host -smp cpus=32 -device virtio-rng-pci -net nic,model=virtio-net-pci -net user,hostfwd=tcp::22231-:22 -drive if=virtio,file=/scratch/local2/ubuntu-jammy-bigbluebutton.qcow2 -nographic -kernel /scratch/local2/vmlinuz-5.15.0-1057-kvm -append "root=/dev/vda3 debug"
    […]
    Run /sbin/init as init process
      with arguments:
        /sbin/init
      with environment:
        HOME=/
        TERM=linux
    Run /etc/init as init process
      with arguments:
        /etc/init
      with environment:
        HOME=/
        TERM=linux
    Run /bin/init as init process
      with arguments:
        /bin/init
      with environment:
        HOME=/
        TERM=linux
    Run /bin/sh as init process
      with arguments:
        /bin/sh
      with environment:
        HOME=/
        TERM=linux
    /bin/sh: 0: can't access tty; job control turned off

Using `-append "root=/dev/vda3 debug init=/usr/bin/systemd"`, it boots to the login prompt.

**Reply by daandemeyer** [reply]
@paulmenzel Install systemd-sysv
